File: content/snippets/go/interview_lcci/04/08_first_common_ancestor.go
package main

import "fmt"

type Node struct {
	Val         int
	Left, Right *Node
}

func lca(root *Node, p int, q int) *Node {
	if root == nil || root.Val == p || root.Val == q {
		return root
	}
	left := lca(root.Left, p, q)
	right := lca(root.Right, p, q)
	if left != nil && right != nil {
		return root
	}
	if left != nil {
		return left
	}
	return right
}

func main() {
	root := &Node{Val: 3}
	root.Left = &Node{Val: 5, Left: &Node{Val: 6}, Right: &Node{Val: 2}}
	root.Right = &Node{Val: 1, Left: &Node{Val: 0}, Right: &Node{Val: 8}}

	fmt.Println(lca(root, 6, 2).Val)
	fmt.Println(lca(root, 6, 8).Val)
}

File: content/snippets/typescript/interview_lcci/04/08_first_common_ancestor.ts
class Node {
  constructor(
    public val: number,
    public left: Node | null = null,
    public right: Node | null = null
  ) {}
}

function lca(root: Node | null, p: number, q: number): Node | null {
  if (!root || root.val === p || root.val === q) return root;
  const left = lca(root.left, p, q);
  const right = lca(root.right, p, q);
  if (left && right) return root;
  return left ?? right;
}

const root = new Node(
  3,
  new Node(5, new Node(6), new Node(2)),
  new Node(1, new Node(0), new Node(8))
);

console.log(lca(root, 6, 2)?.val ?? -1);
console.log(lca(root, 6, 8)?.val ?? -1);
